src/definitions/subscriptionType.ts

```diff
@@ -1,63 +1,14 @@
 import { GraphQLResolveInfo } from 'graphql'
-import {
-  ArgsValue,
-  FieldResolver,
-  GetGen,
-  HasGen3,
-  MaybePromise,
-  MaybePromiseDeep,
-  NeedsResolver,
-  ResultValue,
-} from '../typegenTypeHelpers'
-import { CommonOutputFieldConfig } from './definitionBlocks'
-import { ObjectDefinitionBlock, ObjectDefinitionBuilder, objectType } from './objectType'
+import { ArgsValue, GetGen, MaybePromise, MaybePromiseDeep, ResultValue } from '../typegenTypeHelpers'
+import { CommonOutputFieldConfig, NexusOutputFieldDef } from './definitionBlocks'
+import { ObjectDefinitionBuilder, objectType } from './objectType'
 import { AllNexusOutputTypeDefs } from './wrapping'
+import { BaseScalars } from './_types'
 
-export interface SubscriptionScalarConfig<TypeName extends string, FieldName extends string, T = any>
-  extends CommonOutputFieldConfig<TypeName, FieldName> {
-  /**
-   * Resolve method for the field
-   */
-  resolve?: FieldResolver<TypeName, FieldName>
-
-  subscribe(
-    root: object,
-    args: ArgsValue<TypeName, FieldName>,
-    ctx: GetGen<'context'>,
-    info: GraphQLResolveInfo
-  ): MaybePromise<AsyncIterator<T>> | MaybePromiseDeep<AsyncIterator<T>>
-}
-
-export type ScalarSubSpread<TypeName extends string, FieldName extends string> = NeedsResolver<
-  TypeName,
-  FieldName
-> extends true
-  ? HasGen3<'argTypes', TypeName, FieldName> extends true
-    ? [ScalarSubConfig<TypeName, FieldName>]
-    : [ScalarSubConfig<TypeName, FieldName>] | [FieldResolver<TypeName, FieldName>]
-  : HasGen3<'argTypes', TypeName, FieldName> extends true
-  ? [ScalarSubConfig<TypeName, FieldName>]
-  : [] | [FieldResolver<TypeName, FieldName>] | [ScalarSubConfig<TypeName, FieldName>]
-
-export type ScalarSubConfig<TypeName extends string, FieldName extends string> = NeedsResolver<
-  TypeName,
-  FieldName
-> extends true
-  ? SubscriptionScalarConfig<TypeName, FieldName> & {
-      resolve: FieldResolver<TypeName, FieldName>
-    }
-  : SubscriptionScalarConfig<TypeName, FieldName>
-
-export interface SubscribeFieldConfig<TypeName extends string, FieldName extends string, T = any>
-  extends CommonOutputFieldConfig<TypeName, FieldName> {
-  type: GetGen<'allOutputTypes'> | AllNexusOutputTypeDefs
-
-  /**
-   * Resolve method for the field
-   */
+export interface SubscribeFieldConfigBase<FieldName extends string, Event = any> {
   resolve(
-    root: T,
-    args: ArgsValue<TypeName, FieldName>,
+    root: Event,
```

Comment:
We have some feature work to do to make this typesafe in practice for users.

Reply:
Isn't it typesafe already?

Reply:
No, the resolve `root` type should match the type returned from `subscribe`.

https://github.com/graphql-nexus/schema/issues/454

Reply:
Yeah but the `Event` generic already does that, although it seems it's not properly captured as explained by that issue 👍